### PR TITLE
kubemark use remote cri

### LINKED ISSUE #41504: kubemark should replace the fake docker client with a fake CRI client
Kubelet has switched to using the CRI implementation by default, so CRI is the right abstraction for kubemark to use. This will also ensure all unnecessary functions in dockershim (e.g., creating checkpoint files) are properly mocked out.

The CRI fakes are currently in https://github.com/kubernetes/kubernetes/tree/v1.6.0-alpha.2/pkg/kubelet/api/testing
It'll need some work to match what `FakeDockerClient` does today, as well as some plumbing to use this in the hollow node.

/cc @wojtek-t @gmarek @Random-Liu

### PR DESCRIPTION
**Pull request by gongguan**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes [#41504](https://github.com/kubernetes/kubernetes/issues/41504) 

**Special notes for your reviewer**:
pr [#85879](https://github.com/kubernetes/kubernetes/pull/85879) was reverted by [#86425](https://github.com/kubernetes/kubernetes/pull/86425)
Second commit is added for fix, first is the same as reverted pr.
**Does this PR introduce a user-facing change?**:
```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by mm4tt** [comment]
Looks like kubemark's hollow nodes don't go up with this change. 
An [example hollow kubelet log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/86575/pull-kubernetes-kubemark-e2e-gce-big/1209325666047102976/artifacts/e2e-86575-ac87c-minion-group-1d92/kubelet-hollow-node-25gxx.log) is full of the following errors:
```
E1224 05:14:02.495210       1 kubelet.go:1851] skipping pod synchronization - [container runtime status check may not have completed yet, Container runtime not ready: <nil>]
E1224 05:14:06.225545       1 kubelet.go:2190] Container runtime network not ready: <nil>
E1224 05:14:06.225603       1 kubelet.go:2201] Container runtime not ready: <nil>
```

**Comment by gongguan** [comment]
> Looks like kubemark's hollow nodes don't go up with this change.
> An [example hollow kubelet log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/86575/pull-kubernetes-kubemark-e2e-gce-big/1209325666047102976/artifacts/e2e-86575-ac87c-minion-group-1d92/kubelet-hollow-node-25gxx.log) is full of the following errors:
> 
> ```
> E1224 05:14:02.495210       1 kubelet.go:1851] skipping pod synchronization - [container runtime status check may not have completed yet, Container runtime not ready: <nil>]
> E1224 05:14:06.225545       1 kubelet.go:2190] Container runtime network not ready: <nil>
> E1224 05:14:06.225603       1 kubelet.go:2201] Container runtime not ready: <nil>
> ```

Thanks, there is something wrong with hollow node, I'll fix it.

**Review by mattjmcnaughton — COMMENTED** [review]
@gongguan thanks for tackling this fix. Would you mind summarizing what has changed between this diff and the initial diff which had to be reverted? It'll make it easier for reviewers to know what we should focus on :)

**Comment by gongguan** [comment]
> @gongguan thanks for tackling this fix. Would you mind summarizing what has changed between this diff and the initial diff which had to be reverted? It'll make it easier for reviewers to know what we should focus on :)

Thanks for your kind reminding. The fourth commit which called fix is new, others are the same as reverted pr. There is still something wrong in fake CRI initialization. I'd like to explain it to reviewers when it is ok to review. Help from node reviewers is welcomed now, because I haven't find a complete example of fake CRI usage in existing code.

**Comment by gongguan** [comment]
@mm4tt kubemark-up seems succeed, but there are other failures in pull-kubernetes-kubemark-e2e-gce-big. Do you know whether it's a flaky test or failure of kubemark.

**Comment by gongguan** [comment]
fake remote runtime didn't work normally, ping @feiskyer could you leave some suggestions?

File: staging/src/k8s.io/cri-api/pkg/apis/testing/fake_runtime_service.go [hunk 1]
```diff
@@ -194,6 +194,11 @@ func (r *FakeRuntimeService) RunPodSandbox(config *runtimeapi.PodSandboxConfig,
 			Network: &runtimeapi.PodSandboxNetworkStatus{
 				Ip: FakePodSandboxIPs[0],
 			},
+			Linux: &runtimeapi.LinuxPodSandboxStatus{
```
Comment (gongguan) [inline staging/src/k8s.io/cri-api/pkg/apis/testing/fake_runtime_service.go]: [check](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L445) will always restart sandbox without this

**Comment by mattjmcnaughton** [comment]
@gongguan in case its helpful :) `/retest` will trigger all of the tests, whereas `/test TEST_NAME` will just trigger a single test. I find the latter to be helpful when there's a single test failing, as running just the single test will be quicker. (And as an additional bonus, it prevents CI resources from being used on tests which are already passing).

**Comment by gongguan** [comment]
ping @wojtek-t  @yujuhong for review
second commit is fix for [#85879](https://github.com/kubernetes/kubernetes/pull/85879)

File: staging/src/k8s.io/cri-api/pkg/apis/testing/fake_runtime_service.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/cri-api/pkg/apis/testing/fake_runtime_service.go]: Please add a comment about that then :)

File: pkg/kubelet/remote/fake/fake_runtime.go [hunk 2]
```diff
@@ -53,6 +53,16 @@ func NewFakeRemoteRuntime() *RemoteRuntime {
 	return f
 }
 
+// SetStatus set fake remote runtime's network and runtime condition.
+func (f *RemoteRuntime) SetStatus() {
```
Comment (wojtek-t) [inline pkg/kubelet/remote/fake/fake_runtime.go]: Can we set it as part of Start() method?
Given that for now noone is actually overwriting it anywhere, this seems enough to me.

nit: you may also consider moving the changes to fake_runtime to a separate PR.

Comment (gongguan) [reply pkg/kubelet/remote/fake/fake_runtime.go]: > Can we set it as part of Start() method?
> Given that for now noone is actually overwriting it anywhere, this seems enough to me.

ok
> nit: you may also consider moving the changes to fake_runtime to a separate PR.

Do you mean a pre-pr of this one which modify fake_runtime.go and fake_runtime_service.go?

Comment (wojtek-t) [reply pkg/kubelet/remote/fake/fake_runtime.go]: Yes (those seem to be universal fixes needed independently from this PR).

Comment (gongguan) [reply pkg/kubelet/remote/fake/fake_runtime.go]: done

**Comment by gongguan** [comment]
I've separated fix of fake remote CRI as an independent pr [#86724](https://github.com/kubernetes/kubernetes/pull/86724).
Once it is merged, I'll rebase latest code.

**Comment by gongguan** [comment]
> second commit is fix for [#85879](https://github.com/kubernetes/kubernetes/pull/85879)

As fake CRI fixed, kind ping @wojtek-t @dashpole for review again
